### PR TITLE
SAMZA-2693: Make Samza log4j appenders agnostic of where they are running

### PR DESCRIPTION
**Pull request by cameronlee314**
Issues:
`StreamAppender` is coupled with Samza being run in YARN:
1. `StreamAppender` specifically checks for `isApplicationMaster`, which is a YARN-only term, and it has different initialization logic if it is or isn't in the application master.
2. `StreamAppender` is coupled with `JobModelManager`, which is only used in YARN.
3. `StreamAppender` can only get configs from a specific HTTP coordinator endpoint, so it can't be used in other setups.
For running in Kubernetes (or other environments in general), this class can be generalized better.

Changes:
1. Add a `LoggingContextHolder` which provides general access to config for log appenders regardless of container type. This is still a static class, similar to how appenders were already using the static `JobModelManager.currentJobModelManager`. `StreamAppender` gets config from this `LoggingContextHolder`; it no longer reads the config from `JobModelManager` or an HTTP coordinator endpoint. `JobModelManager` and `ContainerLaunchUtil` were updated to set the config in the `LoggingContextHolder`.
2. Update `StreamAppender` (log4j, log4j2) to have a consistent flow for checking when it can initialize itself, instead of checking application master vs. worker.
3. Refactoring included a bug fix in which the first log event after system initialization was not sent to the log stream.
4. `StreamAppender.getPartitionCount` only returns the overridden partition count value instead of falling back to "job.container.count". This was done to prevent access to the config before it was available, and this is more consistent with the log4j getter pattern for configured values.

Tests:
1. Updated unit tests
2. Ran integration tests

API/usage changes:
1. Constructor for `org.apache.samza.logging.log4j2.StreamAppender` has an extra argument for `LoggingContextHolder`. Any classes which extend this need to be updated to pass the extra argument. Creation of `org.apache.samza.logging.log4j2.StreamAppender` in `log4j2.xml` does not change.
2. `StreamAppender.getPartitionCount` (for both log4j and log4j2) does not fall back to reading "job.container.count". Any classes which use that method and rely on "job.container.count" as a fallback need to be updated to read that config on their own.
3. In worker containers, logs that happen before the call to `LoggingContextHolder.INSTANCE.setConfig` in `ContainerLaunchUtil` will no longer be handled by `StreamAppender`. There are few logs that are impacted by this.

### CODE REVIEW

**Comment by lakshmi-manasa-g** [comment]
why were log4j2 tests deleted?

**Comment by cameronlee314** [comment]
> why were log4j2 tests deleted?

Could you please clarify which deleted tests you are referring to? I don't think there should be anything deleted (just refactored).

**Comment by lakshmi-manasa-g** [comment]
> > why were log4j2 tests deleted?
> 
> Could you please clarify which deleted tests you are referring to? I don't think there should be anything deleted (just refactored).

sorry my bad. found it. missed it cause of large diff follwing deleted file

File: samza-log4j2/src/main/java/org/apache/samza/logging/log4j2/StreamAppender.java [hunk 1]
```diff
@@ -490,6 +455,17 @@ protected void setSerde(Log4jSystemConfig log4jSystemConfig, String systemName)
     }
   }
 
+  /**
+   * If the partition count was explicitly specified, then use that. Otherwise, use the container count as the partition
+   * count.
+   */
+  private int calculateStreamPartitionCount(Config config) {
```
[the remark this answers is not in the document]
Comment (cameronlee314) [reply samza-log4j2/src/main/java/org/apache/samza/logging/log4j2/StreamAppender.java:462]: `getPartitionCount` assumes that the instance variable `config` is available, but `config` isn't actually guaranteed to be available until `LoggingContextHolder` has a config filled in, so this could result in the caller accidentally using `getPartitionCount` too early. `calculateStreamPartitionCount` accepts `config` as an argument, which forces the callers of that method to ensure `config` is available.
I didn't want to change the `getPartitionCount` signature just in case log4j/log4j2 still needed the public getter. Also, based on the javadoc for `getPartitionCount`, it is tied to some log4j/log4j2 configuration, so it seemed like it would be reasonable to have it only consider the configured value.

File: samza-log4j2/src/main/java/org/apache/samza/logging/log4j2/StreamAppender.java [hunk 2]
```diff
@@ -89,38 +78,40 @@
...
+  protected volatile boolean systemInitialized = false;
   protected StreamAppenderMetrics metrics;
   protected long queueTimeoutS = DEFAULT_QUEUE_TIMEOUT_S;
 
+  /**
+   * Constructor is protected so that this class can be extended.
+   */
   protected StreamAppender(String name, Filter filter, Layout<? extends Serializable> layout, boolean ignoreExceptions,
-      boolean usingAsyncLogger, String streamName) {
+      boolean usingAsyncLogger, String streamName, LoggingContextHolder loggingContextHolder) {
```
[the remark this answers is not in the document]
Comment (lakshmi-manasa-g) [reply samza-log4j2/src/main/java/org/apache/samza/logging/log4j2/StreamAppender.java:103]: ah i see. thanks for explaining why not log4j and the need for this extra argument.
however, we are making a breaking change (needing all extended classes to update) for making testing easier. 
i feel we could add another constructor for testing and avoid breaking changes.

File: samza-log4j2/src/main/java/org/apache/samza/logging/log4j2/StreamAppender.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (lakshmi-manasa-g) [reply samza-log4j2/src/main/java/org/apache/samza/logging/log4j2/StreamAppender.java:462]: since this is also a breaking change, i want to dig a bit more into this.
so earlier `getPartitioncount` used getConfig() which fetched the config when config became available (different conditions for AM vs container). So even with this earlier flow, if getPartitionCount was ever called before config became available then there would be a problem. caller of getPartitionCount has already ensured that this call is made only after config becomes available. 

i am seeing LoggingContextHolder as a way to avoid the separate checks for AM and container and even avoid calling them isAM and make it cleaner. it is not going to make config available sooner than before. given this, its more a refactor for cleaner and readable code than logic changes. Hence, i am uncomfortable making breaking changes and would like to avoid if possible.

**Review by lakshmi-manasa-g — COMMENTED** [review]
also for - " logs that happen before the call to LoggingContextHolder.INSTANCE.setConfig in ContainerLaunchUtil will no longer be handled by StreamAppender. There are few logs that are impacted by this."
are there logs generated within samza that will be not sent to StreamAppender?

File: samza-log4j2/src/main/java/org/apache/samza/logging/log4j2/StreamAppender.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (cameronlee314) [reply samza-log4j2/src/main/java/org/apache/samza/logging/log4j2/StreamAppender.java:103]: Good point. I was thinking that unit tests would need to be updated for extenders of this class, so might as well require extenders to update the constructor too, but it makes more sense to leave a backwards compat constructor. I will update this.

File: samza-log4j2/src/main/java/org/apache/samza/logging/log4j2/StreamAppender.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (cameronlee314) [reply samza-log4j2/src/main/java/org/apache/samza/logging/log4j2/StreamAppender.java:462]: Good point that callers of `getPartitionCount` already needed to make sure config was available. I will change this back.

**Comment by cameronlee314** [comment]
> also for - " logs that happen before the call to LoggingContextHolder.INSTANCE.setConfig in ContainerLaunchUtil will no longer be handled by StreamAppender. There are few logs that are impacted by this."
> are there logs generated within samza that will be not sent to StreamAppender?

Yes, `LocalContainerRunner` has a couple logs that would be missed by `StreamAppender`. Also, if there is an external caller of `ContainerLaunchUtil.run` (e.g. Beam), then any logs from that external caller which happen before `ContainerLaunchUtil.run` is called will be missed. For `LocalContainerRunner`, I can move those logs to be after  `ContainerLaunchUtil.run`. For external callers, they can call `LoggingContextHolder.INSTANCE.setConfig` explicitly if they need logging to be available sooner. I will update this in the PR description.

File: samza-core/src/main/java/org/apache/samza/runtime/LocalContainerRunner.java [hunk 3]
```diff
@@ -50,18 +51,22 @@ public static void main(String[] args) throws Exception {
...
-    log.info(String.format("Got coordinator URL: %s", coordinatorUrl));
     System.out.println(String.format("Coordinator URL: %s", coordinatorUrl));
 
     Optional<String> execEnvContainerId = Optional.ofNullable(System.getenv(ShellCommandConfig.ENV_EXECUTION_ENV_CONTAINER_ID));
 
     int delay = new Random().nextInt(SamzaContainer.DEFAULT_READ_JOBMODEL_DELAY_MS()) + 1;
     JobModel jobModel = SamzaContainer.readJobModel(coordinatorUrl, delay);
     Config config = jobModel.getConfig();
+
+    // this call is also in LocalContainerRunner, but adding this here allows more logs to get handled by Samza loggers
```
Comment (lakshmi-manasa-g) [inline samza-core/src/main/java/org/apache/samza/runtime/LocalContainerRunner.java:65]: nit: was this meant to be "this call is also in ContainerLaunchUtil"

File: samza-log4j/src/main/java/org/apache/samza/logging/log4j/StreamAppender.java [hunk 4]
```diff
@@ -408,4 +393,15 @@ private void setSerde(Log4jSystemConfig log4jSystemConfig, String systemName, St
   public Serde<LoggingEvent> getSerde() {
     return serde;
   }
+
+  /**
+   * If the partition count was explicitly specified, then use that. Otherwise, use the container count as the partition
+   * count.
+   */
+  private int calculateStreamPartitionCount(Config config) {
```
Comment (lakshmi-manasa-g) [inline samza-log4j/src/main/java/org/apache/samza/logging/log4j/StreamAppender.java:401]: nit: since getPartitionCount is restored, we dont need this right

**Review by lakshmi-manasa-g — APPROVED** [review]
thank you for making the changes.
lgtm.

File: samza-core/src/main/java/org/apache/samza/runtime/LocalContainerRunner.java (hunk 3 above)
Comment (cameronlee314) [reply samza-core/src/main/java/org/apache/samza/runtime/LocalContainerRunner.java:65]: Thanks!

File: samza-log4j/src/main/java/org/apache/samza/logging/log4j/StreamAppender.java (hunk 4 above)
Comment (cameronlee314) [reply samza-log4j/src/main/java/org/apache/samza/logging/log4j/StreamAppender.java:401]: Nice catch!

File: samza-core/src/main/java/org/apache/samza/logging/LoggingContextHolder.java [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+ */
+public class LoggingContextHolder {
+  private static final Logger LOG = LoggerFactory.getLogger(LoggingContextHolder.class);
+  public static final LoggingContextHolder INSTANCE = new LoggingContextHolder();
+
+  private final AtomicReference<Config> config = new AtomicReference<>();
+
+  @VisibleForTesting
+  LoggingContextHolder() {
+  }
```
Comment (mynameborat) [inline samza-core/src/main/java/org/apache/samza/logging/LoggingContextHolder.java:40]: If the intent is to only use the singleton, this pattern would still let classes under same package break the assumption. 

Thoughts on having reset/clear method if you want to reuse the instance in the test? Not sure which option w/ tradeoffs fits better here.

File: samza-log4j2/src/main/java/org/apache/samza/logging/log4j2/StreamAppender.java [hunk 6]
```diff
@@ -152,18 +144,19 @@ public String getStreamName() {
   }
 
   /**
-   * Getter for the Config parameter.
+   * This should only be called after verifying that the {@link LoggingContextHolder} has the config.
```
Comment (mynameborat) [inline samza-log4j2/src/main/java/org/apache/samza/logging/log4j2/StreamAppender.java:147]: How do we ensure/enforce that?

**Review by mynameborat — COMMENTED** [review]
Few questions. Looks good otherwise

File: samza-log4j2/src/main/java/org/apache/samza/logging/log4j2/StreamAppender.java (hunk 6 above)
Comment (lakshmi-manasa-g) [reply samza-log4j2/src/main/java/org/apache/samza/logging/log4j2/StreamAppender.java:147]: there is actually no tangible way to enforce it. Even prior to this PR, there were only checks to see if the AM's config is available via "JobModelManager.currentJobModelManager() != null" before using the config or calling this getConfig() and no real enforcement. But at that time, there was no explicit comment/doc stating this requirement/caution. now the same thing happens but via "this.loggingContextHolder.getConfig() != null".

the thing about StreamAppender is that because its a log appender but depends on non-logging components' flow and availability (for config..) its flow becomes inter-twined with those components  and hence hard to follow. For example, config get fetched after some component is up and config is asked for within setupSystem which sits as a no-op until config is avail. in other words, the internal methods within this class are dependent and called by other methods BASED on external components flow. This is what lends to this confusing pattern.

File: samza-core/src/main/java/org/apache/samza/logging/LoggingContextHolder.java (hunk 5 above)
Comment (cameronlee314) [reply samza-core/src/main/java/org/apache/samza/logging/LoggingContextHolder.java:40]: Whenever possible, I prefer to avoid using statics in tests. If the statics are not properly reset (and it's not always obvious that a static needs to be reset unless you look at the class), then other tests could be impacted, and that can lead to hard-to-debug failures. For example, a test using a static might pass when run on its own, but then if it runs as part of a full build, it might fail since the static wasn't reset properly from some other test.
I don't think it is too much of a problem that we can't exactly strictly enforce the expected usage. Package-private gets us pretty far, and even if someone abuses the package privacy and uses their own `LoggingContextHolder`, there isn't really anything useful to get out of that. Also, `@VisibleForTesting` documents this as "test only".

File: samza-log4j2/src/main/java/org/apache/samza/logging/log4j2/StreamAppender.java (hunk 6 above)
Comment (cameronlee314) [reply samza-log4j2/src/main/java/org/apache/samza/logging/log4j2/StreamAppender.java:147]: This usage pattern constraint actually already exists in the current implementation. In the AM, it is already not guaranteed that the config is available. I was just adding some documentation to try to help clarify that now.
Unfortunately, to improve this, we would need to change the semantics of some other parts of this class. I actually did make some other changes in the initial iteration, but @lakshmi-manasa-g made a good point about maintaining compatibility within the scope of this PR. There is more discussion of this at https://github.com/apache/samza/pull/1532#discussion_r714217467.
Ideally, there would have been one fetch to the config and then that would have been passed around as a local argument instead of through an instance variable. That would have allowed better enforcement of the availability of the config. For now, this continues to need to be manually checked for ordering of method calls.

Comment (mynameborat) [reply samza-log4j2/src/main/java/org/apache/samza/logging/log4j2/StreamAppender.java:147]: Thanks for the details.

**Review by mynameborat — APPROVED** [review]

File: samza-core/src/main/java/org/apache/samza/runtime/LocalContainerRunner.java [hunk 7]
```diff
@@ -50,18 +51,22 @@ public static void main(String[] args) throws Exception {
...
     Optional<String> execEnvContainerId = Optional.ofNullable(System.getenv(ShellCommandConfig.ENV_EXECUTION_ENV_CONTAINER_ID));
 
     int delay = new Random().nextInt(SamzaContainer.DEFAULT_READ_JOBMODEL_DELAY_MS()) + 1;
     JobModel jobModel = SamzaContainer.readJobModel(coordinatorUrl, delay);
     Config config = jobModel.getConfig();
+
+    // this call is also in ContainerLaunchUtil, but adding this here allows more logs to get handled by Samza loggers
+    LoggingContextHolder.INSTANCE.setConfig(config);
+    log.info(String.format("Got container ID: %s", containerId));
+    log.info(String.format("Got coordinator URL: %s", coordinatorUrl));
```
Comment (alnzng) [inline samza-core/src/main/java/org/apache/samza/runtime/LocalContainerRunner.java:68]: I think leverage log4j format is a better way than using String format:
```java
log.info("Got container ID: {}", containerId);
log.info("Got coordinator URL: {}, coordinatorUrl));
```

**Review by alnzng — COMMENTED** [review]
Overall, looks good to me.

File: samza-core/src/main/java/org/apache/samza/runtime/LocalContainerRunner.java (hunk 7 above)
Comment (cameronlee314) [reply samza-core/src/main/java/org/apache/samza/runtime/LocalContainerRunner.java:68]: Agreed, but I was just moving a few lines around here, so would prefer to not change in this PR.
